packages/contracts/src/contracts/current/protocol/Exchange/MixinExchangeCore.sol

```diff
@@ -76,77 +76,148 @@ contract MixinExchangeCore is
+                return;
+            }
+            
+            // If order.makerTokenAmount is zero, we also reject the order.
+            // While the Exchange contract handles them correctly, they create
+            // edge cases in the supporting infrastructure because they have
+            // an 'infinite' price when computed by a simple division.
+            if(order.makerTokenAmount == 0) {
+                status = uint8(Errors.ORDER_INVALID);
+                return;
+            }
+            
+            // Maker signature needs to be valid
+            if(!isValidSignature(orderHash, order.makerAddress, signature)) {
+                status = uint8(Errors.ORDER_SIGNATURE_INVALID);
+                return;
+            }
         }
-        require(takerTokenFillAmount > 0);
-
+        
         // Validate order expiration
         if (block.timestamp >= order.expirationTimeSeconds) {
-            LogError(uint8(Errors.ORDER_EXPIRED), orderHash);
-            return 0;
+            status = uint8(Errors.ORDER_EXPIRED);
+            return;
         }
-
+        
         // Validate order availability
-        uint256 remainingTakerTokenAmount = safeSub(order.takerTokenAmount, filled[orderHash]);
-        if (remainingTakerTokenAmount == 0) {
-            LogError(uint8(Errors.ORDER_FULLY_FILLED), orderHash);
-            return 0;
+        if (filledAmount >= order.takerTokenAmount) {
+            status = uint8(Errors.ORDER_FULLY_FILLED);
+            return;
         }
-
-        // Validate fill order rounding
-        takerTokenFilledAmount = min256(takerTokenFillAmount, remainingTakerTokenAmount);
-        if (isRoundingError(takerTokenFilledAmount, order.takerTokenAmount, order.makerTokenAmount)) {
-            LogError(uint8(Errors.ROUNDING_ERROR_TOO_LARGE), orderHash);
-            return 0;
+        
+        // Check if order has been cancelled
+        if (cancelled[orderHash]) {
+            status = uint8(Errors.ORDER_CANCELLED);
+            return;
         }
-
+        
         // Validate order is not cancelled
-        if (order.salt < makerEpoch[order.makerAddress]) {
-            LogError(uint8(Errors.ORDER_FULLY_FILLED_OR_CANCELLED), orderHash);
-            return 0;
+        if (makerEpoch[order.makerAddress] > order.salt) {
+            status = uint8(Errors.ORDER_CANCELLED);
+            return;
         }
-
+        
+        // Order is OK
+        status = uint8(Errors.SUCCESS);
+        return;
+    }
+    
+    function getFillAmounts(
+        Order memory order,
+        uint256 filledAmount,
+        uint256 takerTokenFillAmount,
+        address taker)
+        public
+        pure
+        returns (
+            uint8 status,
+            uint256 makerTokenFilledAmount,
+            uint256 takerTokenFilledAmount,
+            uint256 makerFeeAmountPaid,
+            uint256 takerFeeAmountPaid)
+    {
+        // Validate taker
+        if (order.takerAddress != address(0)) {
```

Comment:
This validation seems a bit out of place in this function. It makes more sense to me to either keep it in `fillOrder` or move it to `orderStatus`.